cdk/PgBouncer.ts

```diff
@@ -0,0 +1,179 @@
+import { Construct } from "constructs";
+
+import * as fs from "fs";
+import * as path from "path";
+
+export interface PgBouncerProps {
+  /**
+   * Name for the pgbouncer instance
+   */
+  instanceName: string;
+
+  /**
+   * VPC to deploy PgBouncer into
+   */
+  vpc: ec2.IVpc;
+
+  /**
+   * The RDS instance to connect to
+   */
+  database: {
+    connections: ec2.Connections;
+    secret: secretsmanager.ISecret;
+  };
+
+  /**
+   * Whether to deploy in public subnet
+   * @default false
+   */
+  usePublicSubnet?: boolean;
+
+  /**
+   * Instance type for PgBouncer
+   * @default t3.small
+   */
+  instanceType?: ec2.InstanceType;
+
+  /**
+   * PgBouncer configuration options
+   */
+  pgBouncerConfig: {
+    poolMode: "transaction" | "session" | "statement";
+    maxClientConn: number;
+    defaultPoolSize: number;
+    minPoolSize: number;
+    reservePoolSize: number;
+    reservePoolTimeout: number;
+    maxDbConnections: number;
+    maxUserConnections: number;
+  };
+}
+
+export class PgBouncer extends Construct {
+  public readonly instance: ec2.Instance;
+  public readonly endpoint: string;
+
+  constructor(scope: Construct, id: string, props: PgBouncerProps) {
+    super(scope, id);
+
+    const {
+      vpc,
+      database,
+      usePublicSubnet = false,
+      instanceType = ec2.InstanceType.of(
+        ec2.InstanceClass.T3,
+        ec2.InstanceSize.MICRO,
+      ),
+      pgBouncerConfig = {
+        poolMode: "transaction",
+        maxClientConn: 200,
+        defaultPoolSize: 5,
+        minPoolSize: 0,
+        reservePoolSize: 5,
+        reservePoolTimeout: 5,
+        maxDbConnections: 40,
+        maxUserConnections: 40,
+      },
+    } = props;
+
+    // Create role for PgBouncer instance to enable writing to CloudWatch
+    const role = new iam.Role(this, "InstanceRole", {
+      assumedBy: new iam.ServicePrincipal("ec2.amazonaws.com"),
+      managedPolicies: [
+        iam.ManagedPolicy.fromAwsManagedPolicyName(
+          "AmazonSSMManagedInstanceCore",
+        ),
+        iam.ManagedPolicy.fromAwsManagedPolicyName(
+          "CloudWatchAgentServerPolicy",
+        ),
+      ],
+    });
+
+    // Add policy to allow reading RDS credentials from Secrets Manager
+    role.addToPolicy(
+      new iam.PolicyStatement({
+        actions: ["secretsmanager:GetSecretValue"],
+        resources: [database.secret.secretArn],
+      }),
+    );
+
+    // Create PgBouncer instance
+    this.instance = new ec2.Instance(this, "Instance", {
+      vpc,
+      vpcSubnets: {
+        subnetType: usePublicSubnet
+          ? ec2.SubnetType.PUBLIC
+          : ec2.SubnetType.PRIVATE_WITH_EGRESS,
+      },
+      instanceType,
+      instanceName: props.instanceName,
+      machineImage: ec2.MachineImage.fromSsmParameter(
+        "/aws/service/canonical/ubuntu/server/jammy/stable/current/amd64/hvm/ebs-gp2/ami-id",
+        { os: ec2.OperatingSystemType.LINUX },
+      ),
+      role,
+      detailedMonitoring: true,
```

Comment:
I haven't seen detailed monitoring in action for EC2 instances before, is the difference primarily the frequency of logs produced or is it the kind of log messages generated? The added costs probably aren't significant the grand scheme of things, but was there a noticeable difference for a small instance?